src/lib/redis/scripts.ts

```diff
@@ -20,10 +38,90 @@ type MarkFinishedScript = {
 };
 
 export type RedisScripts = {
+	recordProgress: RecordProgressScript;
+	recordProgressAppend: RecordProgressAppendScript;
 	recordResult: RecordResultScript;
 	markFinished: MarkFinishedScript;
 };
 
+const recordProgress: RecordProgressScript = defineScript({
+	FIRST_KEY_INDEX: 0, // Needed in clusters: https://github.com/redis/node-redis/issues/2521
+	NUMBER_OF_KEYS: 2,
+	SCRIPT: `
+	local keyMeasurementResults = KEYS[1]
+	local keyMeasurementAwaiting = KEYS[2]
+	local testId = ARGV[1]
+	local keyToValueJson = ARGV[2]
+	local date = ARGV[3]
+
+	local probesAwaiting = redis.call('GET', keyMeasurementAwaiting)
+	if not probesAwaiting then
+		return
+	end
+
+	local keyToValue = cjson.decode(keyToValueJson)
+
+	for key, value in pairs(keyToValue) do
+		redis.call('JSON.SET', keyMeasurementResults, key, value)
+	end
+
+	redis.call('JSON.SET', keyMeasurementResults, '$.updatedAt', date)
+	`,
+	transformArguments (measurementId, testId, keyToValue) {
+		return [
+			// keys
+			`gp:m:{${measurementId}}:results`,
+			`gp:m:{${measurementId}}:probes_awaiting`,
+			// values
+			testId,
+			JSON.stringify(Object.fromEntries(Object.entries(keyToValue).map(([ key, value ]) => [ `$.results[${testId}].result.${key}`, JSON.stringify(value) ]))),
+			`"${new Date().toISOString()}"`,
+		];
+	},
+	transformReply () {
+		return null;
+	},
+});
+
+const recordProgressAppend: RecordProgressAppendScript = defineScript({
+	FIRST_KEY_INDEX: 0, // Needed in clusters: https://github.com/redis/node-redis/issues/2521
+	NUMBER_OF_KEYS: 2,
+	SCRIPT: `
+	local keyMeasurementResults = KEYS[1]
+	local keyMeasurementAwaiting = KEYS[2]
+	local testId = ARGV[1]
+	local keyToValueJson = ARGV[2]
+	local date = ARGV[3]
+
+	local probesAwaiting = redis.call('GET', keyMeasurementAwaiting)
+	if not probesAwaiting then
+		return
+	end
+
+	local keyToValue = cjson.decode(keyToValueJson)
+
+	for key, value in pairs(keyToValue) do
+		redis.call('JSON.STRAPPEND', keyMeasurementResults, key, value)
+	end
+
+	redis.call('JSON.SET', keyMeasurementResults, '$.updatedAt', date)
+	`,
+	transformArguments (measurementId, testId, keyToValue) {
+		return [
+			// keys
+			`gp:m:{${measurementId}}:results`,
+			`gp:m:{${measurementId}}:probes_awaiting`,
+			// values
+			testId,
+			JSON.stringify(Object.fromEntries(Object.entries(keyToValue).map(([ key, value ]) => [ `$.results[${testId}].result.${key}`, JSON.stringify(value) ]))),
```

Comment:
Considering the update always has only 1 or 2 keys, maybe it would be better to pass each of them separately and have a simple `if` in the Lua script to avoid the JSON reencoding.

Reply:
HTTP type update may have 3 key, `rawHeaders`, `rawBody`, `rawOutput`. But I think the update still makes sense.